### PR TITLE
Chore: Upgrade Storybook to v7

### PR DESCRIPTION
**Pull request by jackw**
**What is this feature?**

Migrates to latest Storybook (v7.0.5). As much as I love storybook this was a bit of a painful upgrade.

- [x] Run `yarn dlx storybook@next upgrade` to "auto migrate"
- [x] Fix yarn PnP dependency warnings
- [x] Clean up yarnrc file
- [x] Fix remaining breaking changes (.e.g types)
- [x] Make sure storybook builds in dev mode
- [x] Make sure storybook builds in prod mode and serves correctly
- [x] Docs render
- [x] Theme switching works in docs and stories
- [x] Rename internal stories from `*.story.internal.tsx` to `*.internal.story.tsx` to prevent conflicts with storybook framework mapper not being able to index internal stories
- [x] Make sure storybook types don't pollute grafana/ui compiled types (thanks Levitate! 🥳 )

Struggles with latest Storybook direction:

- The `<Preview />` component that most of grafana/ui .mdx files rely on is no longer exposed and it seems the Storybook team are pushing for documenting stories rather than [rendering examples of usage](https://github.com/grafana/grafana/blob/jackw/storybook7/packages/grafana-ui/src/components/Button/Button.mdx?plain=1#L14-L26) directly in .mdx files. To circumvent this for now I've patched the `storybook/blocks` package to expose it and opened [an issue](https://github.com/storybookjs/storybook/issues/21975) with the storybook team.
- The storybook team are looking to drop support for the [subcomponents feature](https://github.com/storybookjs/storybook/issues/20782) which we use in a few places. In SB7 the typings appear to be broken. For now I've marked them with `@ts-ignore`.

**Why do we need this feature?**

To keep storybook dependencies up to date.

**Who is this feature for?**

Grafana Developers

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

If you plan to review these changes locally please clear the local storage in your browser. Storybook stores all sorts of information (like the theme) which can clash with changes the team makes in major updates.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by ashharrison90 — COMMENTED** [review]
amazing work @jackw 👏 

just a couple of minor nits/questions. i'm also seeing some incorrect margin on the button docs page (might be elsewhere, but that's the first one i noticed):

image

**Review by JoaoSilvaGrafana — COMMENTED** [review]
Thanks for doing this Jack 👏 

It seems like the font of the docs has changed, I think it is not picking up our `GlobalStyles`, see example:

Before (includes global styles):
Screenshot 2023-04-14 at 13 32 42

After (doesn't include them, font weight is higher):
Screenshot 2023-04-14 at 13 33 50

File: packages/grafana-ui/src/components/TextArea/TextArea.mdx [hunk 1]
```diff
@@ -1,4 +1,5 @@
-import { Props } from '@storybook/addon-docs/blocks';
+import { ArgTypes, Preview } from '@storybook/blocks';
+import { Field } from '../Forms/Field';
 import { TextArea } from './TextArea';
```
[the remark this answers is not in the document]
Comment (jackw) [reply packages/grafana-ui/src/components/TextArea/TextArea.mdx:3]: Ah thanks for spotting this. Should be fixed now!

**Comment by jackw** [comment]
> amazing work @jackw 👏
> 
> just a couple of minor nits/questions. i'm also seeing some incorrect margin on the button docs page (might be elsewhere, but that's the first one i noticed):
> 
> image

@ashharrison90 unfortunately it seems we have a clash between mdx2 and prettier. `P` tags are now introduced by changes in mdx2 which renders new lines as p tags where previously they would be text. Prettier won't let me remove the line breaks which means we can't do the following to fix it:

```tsx
<Button variant="primary" size="sm" style={{ margin: '5px' }}>Small</Button>
```

It looks like the CSS in the button component doesn't like the `_reboot.scss` reset `p` styles that leak into the docs previews. I pushed a fix in https://github.com/grafana/grafana/pull/65943/commits/33e7e8769e4557caea38d23b504bdada998a656e which isn't ideal but not sure how else to handle it given there doesn't seem to be an `mdx2` option for prettier right now. 🤔

**Comment by jackw** [comment]
> Thanks for doing this Jack 👏
> 
> It seems like the font of the docs has changed, I think it is not picking up our `GlobalStyles`, see example:
> 
> Before (includes global styles): Screenshot 2023-04-14 at 13 32 42
> 
> After (doesn't include them, font weight is higher): Screenshot 2023-04-14 at 13 33 50

@JoaoSilvaGrafana could you try clearing your local storage and trying again? It's a funny one, it seems you need to load a story for the theme to take effect. 🤔

**Review by JoaoSilvaGrafana — APPROVED** [review]
> @JoaoSilvaGrafana could you try clearing your local storage and trying again? It's a funny one, it seems you need to load a story for the theme to take effect. 🤔

This is looking good to me now! Even without clearing local storage so maybe it was something that was in local storage at the time and Ive cleaned it in the mean time 🤷 

I can't see any outstanding issues from my side so, lgtm! 👏 🎉

File: packages/grafana-ui/src/components/ConfirmButton/ConfirmButton.story.tsx [hunk 2]
```diff
@@ -15,6 +15,9 @@ const meta: Meta = {
   title: 'Buttons/ConfirmButton',
   component: ConfirmButton,
   decorators: [withCenteredStory],
+  // SB7 has broken subcomponent types due to dropping support for the feature
+  // https://github.com/storybookjs/storybook/issues/20782
+  // @ts-ignore
```
[the remark this answers is not in the document]
Comment (jackw) [reply packages/grafana-ui/src/components/ConfirmButton/ConfirmButton.story.tsx:20]: Yeah, it just prevents the tabs from appearing for the props table in docs. We're kind of misusing it anyway as it's intended for components like `Table` with `Table.Cell`, `Table.Header` etc. Let's stick with this for now.

**Pull request merged**